File: PolymorphicDeserializationExample/Models/Vehicle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace PolymorphicDeserializationExample.Models
{
    public enum VehicleType
    {
        Bicycle,
        Car
    }

    public abstract class Vehicle
    {
        public VehicleType Type { get; private set; }

        public Vehicle(VehicleType type)
        {
            this.Type = type;
        }
    }

    public class Bicycle : Vehicle
    { 
        public int NumberOfGearsBack { get; private set; }
        public int NumberOfGearsFront { get; private set; }

        public Bicycle(int numberOfGearsFront, int numberOfGearsBack) : base(VehicleType.Bicycle)
        {
            this.NumberOfGearsFront = numberOfGearsFront;
            this.NumberOfGearsBack = numberOfGearsBack;
        }
    }

    public enum EngineType
    {
        Petrol,
        Diesel,
        Electric
    }


    public class Car : Vehicle
    {
        public EngineType EngineType { get; private set; }

        public Car(EngineType engineType): base(VehicleType.Car)
        {
            this.EngineType = engineType;
        }
    }


}

File: PolymorphicDeserializationExample/Models/Vehicle.ts



    // $Classes/Enums/Interfaces(filter)[template][separator]
    // filter (optional): Matches the name or full name of the current item. * = match any, wrap in [] to match attributes or prefix with : to match interfaces or base classes.
    // template: The template to repeat for each matched item
    // separator (optional): A separator template that is placed between all templates e.g. $Properties[public $name: $Type][, ]

    // More info: http://frhagn.github.io/Typewriter/

	
    export abstract class Vehicle {
		constructor(public Type: VehicleType){
		}
    }

    
    export class Bicycle extends Vehicle {
        constructor(public NumberOfGearsBack: number, public NumberOfGearsFront: number){
			super(VehicleType.Bicycle);
		}
    }
    export class Car extends Vehicle {
        constructor(public EngineType: EngineType){
			super(VehicleType.Car);
		}
    }

	
	export type VehicleType =
		"Bicycle" |
		"Car"
	
	export const VehicleType = {
		Bicycle: "Bicycle" as VehicleType,
		Car: "Car" as VehicleType
	}

	export type EngineType =
		"Petrol" |
		"Diesel" |
		"Electric"
	
	export const EngineType = {
		Petrol: "Petrol" as EngineType,
		Diesel: "Diesel" as EngineType,
		Electric: "Electric" as EngineType
	}

